### PR TITLE
Sector index

### PR DESCRIPTION
**Pull request by artyom-yurin**
### Description of the Change
Added Sector Index to control storages

### Benefits
Component of StoageMiner

### Possible Drawbacks 
None

### CODE REVIEW

File: core/primitives/sector_file/sector_file.hpp [hunk 1]
```diff
@@ -18,7 +20,21 @@ namespace fc::primitives::sector_file {
...
+  const std::vector<SectorFileType> kSectorFileTypes = {
+      SectorFileType::FTUnsealed,
+      SectorFileType::FTSealed,
+      SectorFileType::FTCache};
+
+  // 10x overheads
+  const std::unordered_map<SectorFileType, int> kOverheadSeal{
+      {SectorFileType::FTUnsealed, 10},
+      {SectorFileType::FTSealed, 10},
+      {SectorFileType::FTCache, 70}  // TODO(artyom-yurin): confirm for 32G
```
Comment (igor-egorov) [inline core/primitives/sector_file/sector_file.hpp:32]: Please create a task in jira and leave the id here.

File: core/markets/storage/example/resources.hpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * Selective car generated from binary payload from go-fil-markets integration
+ * test
+ * https://github.com/filecoin-project/go-fil-markets/blob/5e9a9cbac406599049c4cef6f67527fbc49a9684/storagemarket/integration_test.go#L39
+ */
+#define CAR_FROM_PAYLOAD_FILE "/Users/soramitsu/cpp-filecoin/core/markets/storage/example//payload_selective.car"
```
Comment (igor-egorov) [inline core/markets/storage/example/resources.hpp:11]: Committing a local absolute path seems unacceptable.

File: core/primitives/uri_parser/percent_encoding.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Dmitriy Khaustov aka xDimon
+// Contacts: khaustov.dm@gmail.com
+// File created on: 2017.05.30
+
+// PercentEncoding.cpp
+
+
```
Comment (igor-egorov) [inline core/primitives/uri_parser/percent_encoding.cpp:21]: Please put Soramitsu license header into all to be committed non-binary files too.

File: core/sector_storage/stores/impl/index_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  using std::chrono::system_clock;
+
+  std::string toSectorsID(const SectorId &sector_id,
+                          const SectorFileType &file_type) {
+    return sectorName(sector_id) + "_" + toString(file_type);
+  }
+
+  bool isValidUrl(const std::string &url) {
+    static std::regex url_regex(
+        "https?:\\/\\/"
```
Comment (igor-egorov) [inline core/sector_storage/stores/impl/index_impl.cpp:27]: please leave a reference to the regex source

File: core/sector_storage/stores/impl/index_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        continue;
+      }
+
+      std::string sector_id = toSectorsID(sector, type);
+      auto sector_iter = sectors_.find(sector_id);
+      if (sector_iter == sectors_.end()) {
+        continue;
+      }
+      for (const auto &id : sector_iter->second) {
+        storage_ids[id]++;
```
Comment (igor-egorov) [inline core/sector_storage/stores/impl/index_impl.cpp:174]:
```suggestion
        ++storage_ids[id];
```

File: core/sector_storage/stores/impl/index_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+      }
+      if (!sealing && !storage.info.can_store) {
+        continue;
+      }
+
+      if (req_space > storage.fs_stat.available) {
+        continue;
+      }
+
+      if (system_clock::now() - storage.last_heartbreak
```
Comment (igor-egorov) [inline core/sector_storage/stores/impl/index_impl.cpp:256]: image

should it be a heartbeat?

File: core/sector_storage/stores/index.hpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  using fc::primitives::sector_file::SectorFileType;
+  using std::chrono::system_clock;
+
+  const std::chrono::seconds kHeartbeatInterval(10);
+  const std::chrono::seconds kSkippedHeartbeatThreshold =
+      kHeartbeatInterval * 5;
+
+  struct StorageInfo {
+    ID id;
+    std::vector<std::string> urls;  // TODO: Support non-http transports
```
Comment (igor-egorov) [inline core/sector_storage/stores/index.hpp:29]: Not clear what changes are implied by the todo. Please create a task in jira with the desired changes description and leave the id here.

File: test/core/primitives/sector_file/sector_file_test.cpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+#include <gtest/gtest.h>
+#include "primitives/sector_file/sector_file.hpp"
+
+#include "testutil/outcome.hpp"
+
+using namespace fc::primitives::sector_file;
+
+TEST(SealSpaceUse, Success){
```
Comment (igor-egorov) [inline test/core/primitives/sector_file/sector_file_test.cpp:13]: Please add given/when/then or at least single-line docstring as test description.

File: core/sector_storage/stores/impl/index_impl.cpp (hunk 4 above)
Comment (artyom-yurin) [reply core/sector_storage/stores/impl/index_impl.cpp:27]: https://stackoverflow.com/questions/38608116/how-to-check-a-specified-string-is-a-valid-url-or-not-using-c-code

File: test/core/primitives/sector_file/sector_file_test.cpp (hunk 8 above)
Comment (artyom-yurin) [reply test/core/primitives/sector_file/sector_file_test.cpp:18]: Fixed

File: core/sector_storage/stores/index.hpp (hunk 7 above)
Comment (artyom-yurin) [reply core/sector_storage/stores/index.hpp]: Fixed

File: core/sector_storage/stores/impl/index_impl.cpp (hunk 6 above)
Comment (artyom-yurin) [reply core/sector_storage/stores/impl/index_impl.cpp]: Fixed

File: core/sector_storage/stores/impl/index_impl.cpp (hunk 5 above)
Comment (artyom-yurin) [reply core/sector_storage/stores/impl/index_impl.cpp]: Fixed

File: core/primitives/uri_parser/percent_encoding.cpp (hunk 3 above)
Comment (artyom-yurin) [reply core/primitives/uri_parser/percent_encoding.cpp:26]: Added

File: core/markets/storage/example/resources.hpp (hunk 2 above)
Comment (artyom-yurin) [reply core/markets/storage/example/resources.hpp]: Removed file

File: core/primitives/sector_file/sector_file.hpp (hunk 1 above)
Comment (artyom-yurin) [reply core/primitives/sector_file/sector_file.hpp]: Added

File: core/sector_storage/stores/impl/index_impl.cpp (hunk 6 above)
Comment (Alexey-N-Chernyshov) [reply core/sector_storage/stores/impl/index_impl.cpp]: nope?

File: core/primitives/uri_parser/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Copyright Soramitsu Co., Ltd. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (artyom-yurin) [reply core/primitives/uri_parser/CMakeLists.txt]: Moved

File: core/primitives/sector_file/sector_file.cpp [hunk 10]
```diff
@@ -19,6 +19,25 @@ namespace fc::primitives::sector_file {
     return "<unknown " + std::to_string(file_type) + ">";
   }
 
+  outcome::result<uint64_t> sealSpaceUse(SectorFileType file_type, RegisteredProof seal_proof_type) {
```
[the remark this answers is not in the document]
Comment (artyom-yurin) [reply core/primitives/sector_file/sector_file.cpp]: Done

**Pull request merged**